### PR TITLE
support ascend npu platform.

### PR DESCRIPTION
**Pull request by momo609**
support ascend npu platform.

### CODE REVIEW

File: src/speculators/train/checkpointer.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
     set_optimizer_state_dict,
 )
 from transformers.modeling_utils import PreTrainedModel
+from speculators.utils.util import is_torch_npu_available
```
Comment (dsikka) [inline src/speculators/train/checkpointer.py]: For many of these checks, I think it would be easier to do something like this:

```python 

from transformers.utils import is_torch_npu_available

def fetch_device():
    if is_torch_npu_available():
       return "npu:0"
   else:
       return "cuda:0"
```

And then just update the calls as opposed to repeating lines:

``` python
full_state_dict = load_safetensors_state_dict(
            self.model_path(self.previous_epoch),
            fetch_device()  
 )
```

Comment (fynnsu) [reply src/speculators/train/checkpointer.py]: You could also use the [torch.accelerator](https://docs.pytorch.org/docs/stable/accelerator.html) api which generalizes across different accelerator backends.

e.g. `torch.accelerator.is_available()`

Comment (fynnsu) [reply src/speculators/train/checkpointer.py]: Or to get the device label (type:idx)
```python
device = f"{torch.accelerator.current_accelerator().type}:{torch.accelerator.current_device_index()}"
```
returns `cuda:0` on an nvidia box and should generalize to other hardware types.

File: src/speculators/data_generation/vllm_hidden_states_generator.py [hunk 2]
```diff
@@ -24,6 +24,7 @@
 from vllm.v1.executor.multiproc_executor import MultiprocExecutor
 from vllm.v1.request import Request, RequestStatus
 from vllm.v1.structured_output import StructuredOutputManager
+from speculators.utils.util import is_torch_npu_available
```
Comment (shanjiaz) [inline src/speculators/data_generation/vllm_hidden_states_generator.py]: The imports are failing because the util is called `is_npu_available`

File: src/speculators/data_generation/config_generator.py [hunk 3]
```diff
@@ -30,15 +31,19 @@
...
+    :return: GPU model and count or NPU model and count, or "CPU only" if no GPU/NPU available
     """
-    if not torch.cuda.is_available():
+    if torch.cuda.is_available():
+        gpu_count = torch.cuda.device_count()
+        gpu_name = torch.cuda.get_device_name(0)
+        return gpu_name if gpu_count == 1 else f"{gpu_count}x {gpu_name}"
+    elif is_torch_npu_available():
+        npu_count = torch.npu.device_count()
+        npu_name = torch.cnpuuda.get_device_name(0)
```
Comment (shanjiaz) [inline src/speculators/data_generation/config_generator.py]: Typo. This should probably be `torch.npu.get_device_name`

**Review by shanjiaz — CHANGES_REQUESTED** [review]
Thanks so much for contributing! Would you mind taking another look at the failing tests? It would be great if you could try adding some NPU tests and share the results with us as well. Feel free to reach out with any questions/concerns. You can also find us in vllm slack speculators channel!

**Comment by momo609** [comment]
> Thanks so much for contributing! Would you mind taking another look at the failing tests? It would be great if you could try adding some NPU tests and share the results with us as well. Feel free to reach out with any questions/concerns. You can also find us in vllm slack speculators channel!

I'll adding some test results on NPU. We also have some of our own trained model weights, would it be okay to include a link to these in the repository as well?

**Comment by momo609** [comment]
> > > Thanks so much for contributing! Would you mind taking another look at the failing tests? It would be great if you could try adding some NPU tests and share the results with us as well. Feel free to reach out with any questions/concerns. You can also find us in vllm slack speculators channel!
> > 
> > 
> > I'll adding some test results on NPU. We also have some of our own trained model weights, would it be okay to include a link to these in the repository as well?
> 
> @momo609 Could you take another look at the failed quality tests? You can run `tox -e quality` and `pre-commit run --all-files` locally to check. Thanks!

fixed

File: scripts/gen_and_train.py [hunk 4]
```diff
@@ -163,6 +165,16 @@ def print_block(title: str, content: str):
...
+def npu_command(python_alt, command):
+    if is_npu_available():
+        if python_alt and python_alt.strip():
+            command = []
+            command.extend(python_alt.split())
+        else:
+            command = ["python"]
+    return command
+
+
```
Comment (fynnsu) [inline scripts/gen_and_train.py:181]: It seems like the purpose of this is to remove the use of `uv` to setup the environment and instead just call python directly (using whatever environment is enabled).

The reason we're using `uv` for this is because the environment requirements for the data generation phase are different than for training. The logic in `run_script` allows us to create an ephemeral environment with all the required packages for each command that is run.

If the problem is that you require additional packages (such as to support torch npu), then they can be added to the `requires` list of `run_script` when calling `run_script`.

File: src/speculators/data_generation/vllm_hidden_states_generator.py [hunk 5]
```diff
@@ -310,8 +312,7 @@ def generate(self, token_ids: list[list[int]] | torch.Tensor) -> list[dict]:
                 }
             )
             offset += seq_len
-
-        torch.cuda.empty_cache()
+        empty_cache()
```
Comment (fynnsu) [inline src/speculators/data_generation/vllm_hidden_states_generator.py:315]: There is actually a `torch.accelerator` fn for this one [torch.accelerator.empty_cache()](https://docs.pytorch.org/docs/stable/generated/torch.accelerator.memory.empty_cache.html).

File: src/speculators/train/trainer.py [hunk 6]
```diff
@@ -101,7 +102,10 @@ def setup_model(self):
                 for m in self.model.layers.children():  # type: ignore[union-attr]
                     if not isinstance(m, FSDPModule):
                         continue
-                    m.to_empty(device="cuda")  # type: ignore[attr-defined]
+                    if is_npu_available():
+                        m.to_empty(device="npu")  # type: ignore[attr-defined]
+                    else:
+                        m.to_empty(device="cuda")  # type: ignore[attr-defined]
```
Comment (fynnsu) [inline src/speculators/train/trainer.py]:
```suggestion
                    acc_type = torch.accelerator.current_accelerator().type
                    m.to_empty(device=acc_type)  # type: ignore[attr-defined]
```

File: scripts/gen_and_train.py (hunk 4 above)
Comment (momo609) [reply scripts/gen_and_train.py:181]: > It seems like the purpose of this is to remove the use of `uv` to setup the environment and instead just call python directly (using whatever environment is enabled).
> 
> The reason we're using `uv` for this is because the environment requirements for the data generation phase are different than for training. The logic in `run_script` allows us to create an ephemeral environment with all the required packages for each command that is run.
> 
> If the problem is that you require additional packages (such as to support torch npu), then they can be added to the `requires` list of `run_script` when calling `run_script`.

Yes, but vllm_ascend currently depends on PyTorch 2.8, which conflicts with the latest vLLM dependencies, so we directly launched it with Python and manually installed the required dependencies.

**Comment by momo609** [comment]
I have run 'tox -e quality' and successs，but ci seems don't have this fuc?
image

**Comment by dsikka** [comment]
@momo609 I believe you are seeing type failures, not style. Do you mind running `tox -e quality`?
This should show the failures I think

File: src/speculators/train/trainer.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
     SingleGPUCheckpointer,
 )
 from speculators.train.utils import apply_fully_sharded
+from speculators.utils.util import is_npu_available
```
Comment (rahul-tuli) [inline src/speculators/train/trainer.py]: This function is not used in current file, we need to remove it to make quality pass, kindly refer to the patch I pasted

Comment (momo609) [reply src/speculators/train/trainer.py]: @rahul-tuli  fixed.

Comment (rahul-tuli) [reply src/speculators/train/trainer.py]: Nice, Let's go!

File: scripts/gen_and_train.py [hunk 8]
```diff
@@ -167,6 +169,16 @@ def print_block(title: str, content: str):
     print("\n", "#" * term_width, "\n", sep="")
 
 
+def npu_command(python_alt, command):
```
Comment (dsikka) [inline scripts/gen_and_train.py:172]: nit: this should also go in utils, not in the generic train script

File: scripts/gen_and_train.py [hunk 9]
```diff
@@ -167,6 +169,16 @@ def print_block(title: str, content: str):
...
+    if is_npu_available():
+        if python_alt and python_alt.strip():
+            command = []
+            command.extend(python_alt.split())
+        else:
+            command = ["python"]
+    return command
+
```
Comment (dsikka) [inline scripts/gen_and_train.py:180]: same as above
